cpu/stm32_common/periph/pm.c

```diff
@@ -41,6 +48,176 @@
 #define PM_STOP_CONFIG (PWR_CR_LPDS | PWR_CR_FPDS)
 #endif
 
+#if defined(CPU_FAM_STM32L1)
+/* Variables definitions to LPM */
+static uint32_t ahb_gpio_clocks;
+static uint32_t tmpreg;
+static uint32_t lpm_gpio_moder[CPU_NUMBER_OF_PORTS];
+static uint32_t lpm_gpio_pupdr[CPU_NUMBER_OF_PORTS];
+static uint16_t lpm_gpio_otyper[CPU_NUMBER_OF_PORTS];
+static uint16_t lpm_gpio_odr[CPU_NUMBER_OF_PORTS];
+static uint16_t lpm_portmask_system[CPU_NUMBER_OF_PORTS] = { 0 };
+static uint16_t lpm_portmask_user[CPU_NUMBER_OF_PORTS] = { 0 };
+// static uint8_t  lpm_usart[UART_NUMOF];
+
+static inline GPIO_TypeDef *_port(gpio_t pin)
+{
+    return (GPIO_TypeDef *)(pin & ~(0x0f));
+}
+
+/**
+ * @brief   Extract the port number form the given identifier
+ *
+ * The port number is extracted by looking at bits 10, 11, 12, 13 of the base
+ * register addresses.
+ */
+static inline int _port_num(gpio_t pin)
+{
+    return ((pin >> 10) & 0x0f);
+}
+
+/**
+ * @brief   Extract the pin number from the last 4 bit of the pin identifier
+ */
+static inline int _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+/* We are not using gpio_init as it sets GPIO clock speed to maximum */
+/* We add GPIOs we touched to exclusion mask lpm_portmask_system */
+static void pin_set(gpio_t pin, uint8_t value) {
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
+    tmpreg = port->MODER;
+    tmpreg &= ~(3 << (2*pin_num));
+    tmpreg |= (1 << (2*pin_num));
+    port->MODER = tmpreg;
+
+    port->PUPDR &= ~(3 << (2*pin_num));
+    port->OTYPER &= ~(1 << pin_num);
+    if (value) {
+        port->ODR |= (1 << pin_num);
+    } else {
+        port->ODR &= ~(1 << pin_num);
+    }
+
+    // lpm_portmask_system[((uint32_t)&port>> 10) & 0x0f] |= 1 << pin;
+    lpm_portmask_system[_port_num(pin)] |= 1 << pin_num;
+}
+
+/* Do not change GPIO state in sleep mode */
+void lpm_arch_add_gpio_exclusion(gpio_t gpio) {
+	uint8_t port = ((uint32_t)gpio >> 10) & 0x0f;
+	uint8_t pin = ((uint32_t)gpio & 0x0f);
+
+	lpm_portmask_user[port] |= (uint16_t)(1<<pin);
+}
+
+/* Change GPIO state to AIN in sleep mode */
+void lpm_arch_del_gpio_exclusion(gpio_t gpio) {
+	uint8_t port = ((uint32_t)gpio >> 10) & 0x0f;
+	uint8_t pin = ((uint32_t)gpio & 0x0f);
+
+	lpm_portmask_user[port] &= ~(uint16_t)(1<<pin);
+}
+
+void lpm_before_i_go_to_sleep(void){
+    uint8_t i;
+    uint8_t p;
+    uint32_t mask;
+    GPIO_TypeDef *port;
+
+
+    for (i = 0; i < CPU_NUMBER_OF_PORTS; i++) {
+        port = (GPIO_TypeDef *)(GPIOA_BASE + i*(GPIOB_BASE - GPIOA_BASE));
+
+        /* save GPIO registers values */
+        lpm_gpio_moder[i] = port->MODER;
+        lpm_gpio_pupdr[i] = port->PUPDR;
+        lpm_gpio_otyper[i] = (uint16_t)(port->OTYPER & 0xFFFF);
+        lpm_gpio_odr[i] = (uint16_t)(port->ODR & 0xFFFF);
+	}
```

Comment:
Are you sure you want to put ALL ports in low power? It seems to me a bit overkill and might slow down the process of sleeping. I think it can be better to leave this option to the user who will set the power modes of the peripherals he/she uses.